baremetal/metal3machine_manager.go

```diff
@@ -1801,22 +1904,90 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
 	return nodes, nodesCount, err
 }
 
-// getMatchingNodesWithoutLabelCount tLabel gets kubernetes nodes based on their Spec.providerID field.
-func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context, providerIDLegacy, providerIDNew string, providerIDonM3M *string, clientFactory ClientGetter) (int, error) {
+func (m *MachineManager) SetNodeProviderIDByHostname(ctx context.Context, clientFactory ClientGetter) error {
 	corev1Remote, err := clientFactory(ctx, m.client, m.Cluster)
-	matchingNodesCount := 0
 	if err != nil {
-		return matchingNodesCount, errors.Wrap(err, "Error creating a remote client")
+		return errors.Wrap(err, "Error creating a remote client")
+	}
+
+	metal3MachineHostnames := m.getMetal3MachineHostnames()
+
+	if len(metal3MachineHostnames) == 0 {
+		errMessage := "No machine hostname"
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	nodes, err := corev1Remote.Nodes().List(ctx, metav1.ListOptions{})
+
+	if err != nil {
+		m.Log.Error(err, "error while retrieving nodes")
+		return WithTransientError(err, requeueAfter)
+	}
+
+	if len(nodes.Items) == 0 {
+		errMessage := "no Nodes found"
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	matchingNodes := []corev1.Node{}
+
+	for _, node := range nodes.Items {
+		hostnameLabel, ok := node.GetLabels()["kubernetes.io/hostname"]
+		if !ok {
+			continue
+		}
+
+		for _, metal3MachineHostname := range metal3MachineHostnames {
+			if metal3MachineHostname == hostnameLabel {
+				matchingNodes = append(matchingNodes, node)
+				break
+			}
+		}
+	}
+
+	if len(nodes.Items) == 0 {
+		errMessage := fmt.Sprintf("no nodes matching hostnames: %s", strings.Join(metal3MachineHostnames, ", "))
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	if len(matchingNodes) > 1 {
+		errMessage := fmt.Sprintf("multiple nodes matching hostnames: %s", strings.Join(metal3MachineHostnames, ", "))
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	node := matchingNodes[0]
+
+	m.Log.Info("found one, setting provider id on Node")
```

Comment:
nit: maybe add a node ID to logging? Also I wonder if "found one" will be clear in the context of the entire logs